### PR TITLE
feat: distributed execution of replace into statement

### LINKED ISSUE #11888: Feature: distributed execution of `replace into` statement
**Summary**

- [ ] using query
- [ ] using stage

### PR DESCRIPTION
**Pull request by SkyFan2002**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

- Closes #11888

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/12119)

### CODE REVIEW

**Comment by SkyFan2002** [comment]
askbend:summary

**Comment by JackTan25** [comment]
please add more tests, I think we need an exception test. you can use like "replace into xx ... from xx.parquet",but the fields of parquets don't match the fields num with target table.

File: src/query/service/src/interpreters/interpreter_copy.rs [hunk 1]
```diff
@@ -476,22 +259,16 @@ impl Interpreter for CopyInterpreter {
...
-                            .await?;
-
-                        Ok(build_res)
-                    } else {
-                        self.build_local_copy_into_table_pipeline(plan).await
-                    }
-                } else {
-                    self.build_local_copy_into_table_pipeline(plan).await
-                }
+                let (physical_plan, files) = self.build_physical_plan(plan).await?;
```
Comment (JackTan25) [inline src/query/service/src/interpreters/interpreter_copy.rs:262]: if files are empty, can we do an optimization?

File: src/query/service/src/interpreters/interpreter_copy.rs [hunk 2]
```diff
@@ -177,64 +179,34 @@ impl CopyInterpreter {
...
+            stage_table_info: plan.stage_table_info.clone(),
+            table_info: to_table.get_table_info().clone(),
+            force: plan.force,
+            write_mode: plan.write_mode,
+            validation_mode: plan.validation_mode.clone(),
+
+            files: files.clone(),
+            source,
+        }));
+        if plan.enable_distributed {
```
Comment (JackTan25) [inline src/query/service/src/interpreters/interpreter_copy.rs:201]: if there is only one file, should we need distributed pipeline?

File: src/query/service/src/schedulers/fragments/plan_fragment.rs [hunk 3]
```diff
@@ -194,6 +211,68 @@ impl PlanFragment {
         Ok(fragment_actions)
     }
 
+    fn reshuffle<T: Clone>(
```
Comment (JackTan25) [inline src/query/service/src/schedulers/fragments/plan_fragment.rs:214]: I think we can use this method in src/query/catalog/src/plan/partition.rs:reshuffle to reduce duplicate codes. But we can do this in the next pr.

Comment (SkyFan2002) [reply src/query/service/src/schedulers/fragments/plan_fragment.rs:214]: exception has been coverd in https://github.com/datafuselabs/databend/blob/main/tests/sqllogictests/suites/base/09_fuse_engine/09_0023_replace_into

File: src/query/sql/src/executor/physical_plan.rs [hunk 4]
```diff
@@ -784,66 +789,41 @@ impl UnionAll {
...
-    pub plan_id: u32,
-    pub catalog_info: CatalogInfo,
-    pub database_name: String,
-    pub table_name: String,
-
-    pub required_values_schema: DataSchemaRef, // ... into table(<columns>) ..  -> <columns>
-    pub values_consts: Vec<Scalar>,            // (1, ?, 'a', ?) -> (1, 'a')
-    pub required_source_schema: DataSchemaRef, // (1, ?, 'a', ?) -> (?, ?)
-    // these three fields are used for query result render
+pub struct QueryCtx {
```
Comment (dantengsky) [inline src/query/sql/src/executor/physical_plan.rs:814]: Please consider rename this struct,  too similar to the heavily use struct `QueryContext`...

File: src/query/sql/src/executor/physical_plan.rs [hunk 5]
```diff
@@ -1077,12 +1121,12 @@ impl PhysicalPlan {
             PhysicalPlan::ProjectSet(_) => "Unnest".to_string(),
             PhysicalPlan::RuntimeFilterSource(_) => "RuntimeFilterSource".to_string(),
             PhysicalPlan::DeletePartial(_) => "DeletePartial".to_string(),
-            PhysicalPlan::DeleteFinal(_) => "DeleteFinal".to_string(),
+            PhysicalPlan::MutationAggregate(_) => "DeleteFinal".to_string(),
```
Comment (dantengsky) [inline src/query/sql/src/executor/physical_plan.rs:1124]:
```suggestion
            PhysicalPlan::MutationAggregate(_) => "MutationAggregate".to_string(),
```

File: src/query/storages/fuse/src/operations/replace_into/processors/processor_unbranched_replace_into.rs [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        PipeItem::create(processor_ptr, vec![input], vec![
+            output_port_merge_into_action,
+        ])
+    }
+}
+
+#[async_trait::async_trait]
+impl Processor for UnbranchedReplaceIntoProcessor {
+    fn name(&self) -> String {
+        "ReplaceIntoTransform".to_owned()
```
Comment (dantengsky) [inline src/query/storages/fuse/src/operations/replace_into/processors/processor_unbranched_replace_into.rs:102]:
```suggestion
        "UnbranchedReplaceIntoProcessor".to_owned()
```

**Comment by dantengsky** [comment]
`rr long-run`, 1000 iterations, local fs + standalone mode passed

2023-08-10-195823_1252x995_scrot

**Comment by dantengsky** [comment]
3 nodes, distributed deployment,  local fs, 10000 iteration 

 PASSED
 
 ~~~
MySQL [(none)]> select * from system.clusters;
+------------------------+-----------+-------+--------------------------------------------------------------------------------------------------------------+
| name                   | host      | port  | version                                                                                                      |
+------------------------+-----------+-------+--------------------------------------------------------------------------------------------------------------+
| JQQmH7yG3CjyKSqZY2LyD4 | 127.0.0.1 | 19093 | v1.2.55-nightly-f545f7c2cb08567013c327acfc5271ac3531e274(rust-1.72.0-nightly-2023-08-10T09:27:14.168048062Z) |
| KyVJW8dsE1wYbkiepZrPA4 | 127.0.0.1 | 19091 | v1.2.55-nightly-f545f7c2cb08567013c327acfc5271ac3531e274(rust-1.72.0-nightly-2023-08-10T09:27:14.168048062Z) |
| fYzUEb3BGIZfAcmlvT1FA1 | 127.0.0.1 | 19092 | v1.2.55-nightly-f545f7c2cb08567013c327acfc5271ac3531e274(rust-1.72.0-nightly-2023-08-10T09:27:14.168048062Z) |
+------------------------+-----------+-------+--------------------------------------------------------------------------------------------------------------+
3 rows in set (0.002 sec)

MySQL [(none)]> set global enable_distributed_replace_into = 1;

 ~~~

~~~
[2023-08-10T14:29:59Z INFO  test_replace_recluster] ==========================
[2023-08-10T14:29:59Z INFO  test_replace_recluster] ====verify table state====
[2023-08-10T14:29:59Z INFO  test_replace_recluster] ==========================
[2023-08-10T14:29:59Z INFO  test_replace_recluster]
[2023-08-10T14:29:59Z INFO  test_replace_recluster]
[2023-08-10T14:29:59Z INFO  test_replace_recluster] number of successfully executed replace-into statements : 8326
[2023-08-10T14:29:59Z INFO  test_replace_recluster]
[2023-08-10T14:29:59Z INFO  test_replace_recluster]
[2023-08-10T14:29:59Z INFO  test_replace_recluster] CHECK: value of successfully executed replace into statements
[2023-08-10T14:29:59Z INFO  test_replace_recluster] CHECK: value of correlated column
[2023-08-10T14:29:59Z INFO  test_replace_recluster] CHECK: full table scanning
[2023-08-10T14:30:03Z INFO  test_replace_recluster] ===========================
[2023-08-10T14:30:03Z INFO  test_replace_recluster] ======     PASSED      ====
[2023-08-10T14:30:03Z INFO  test_replace_recluster] ===========================
[2023-08-10T14:30:03Z INFO  test_replace_recluster]
[2023-08-10T14:30:03Z INFO  test_replace_recluster]
[2023-08-10T14:30:03Z INFO  test_replace_recluster] ========METRICS============
[2023-08-10T14:30:03Z INFO  test_replace_recluster] KyVJW8dsE1wYbkiepZrPA4  fuse_commit_mutation_unresolvable_conflict : 8896.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] KyVJW8dsE1wYbkiepZrPA4  fuse_replace_into_block_number_after_pruning : 74331.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] fYzUEb3BGIZfAcmlvT1FA1  fuse_replace_into_block_number_after_pruning : 37121.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] JQQmH7yG3CjyKSqZY2LyD4  fuse_replace_into_block_number_after_pruning : 199039.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] KyVJW8dsE1wYbkiepZrPA4  fuse_replace_into_block_number_bloom_pruned : 70404.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] fYzUEb3BGIZfAcmlvT1FA1  fuse_replace_into_block_number_bloom_pruned : 34972.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] JQQmH7yG3CjyKSqZY2LyD4  fuse_replace_into_block_number_bloom_pruned : 180515.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] KyVJW8dsE1wYbkiepZrPA4  fuse_replace_into_block_number_source : 32221.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] KyVJW8dsE1wYbkiepZrPA4  fuse_replace_into_block_number_totally_loaded : 3541.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] fYzUEb3BGIZfAcmlvT1FA1  fuse_replace_into_block_number_totally_loaded : 2020.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] JQQmH7yG3CjyKSqZY2LyD4  fuse_replace_into_block_number_totally_loaded : 17260.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] KyVJW8dsE1wYbkiepZrPA4  fuse_replace_into_block_number_whole_block_deletion : 53.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] JQQmH7yG3CjyKSqZY2LyD4  fuse_replace_into_block_number_whole_block_deletion : 217.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] KyVJW8dsE1wYbkiepZrPA4  fuse_replace_into_block_number_write : 3541.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] fYzUEb3BGIZfAcmlvT1FA1  fuse_replace_into_block_number_write : 2020.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] JQQmH7yG3CjyKSqZY2LyD4  fuse_replace_into_block_number_write : 17260.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] KyVJW8dsE1wYbkiepZrPA4  fuse_replace_into_block_number_zero_row_deleted : 333.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] fYzUEb3BGIZfAcmlvT1FA1  fuse_replace_into_block_number_zero_row_deleted : 129.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] JQQmH7yG3CjyKSqZY2LyD4  fuse_replace_into_block_number_zero_row_deleted : 1047.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] KyVJW8dsE1wYbkiepZrPA4  fuse_replace_into_number_accumulate_merge_action : 9917.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] fYzUEb3BGIZfAcmlvT1FA1  fuse_replace_into_number_accumulate_merge_action : 3489.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] JQQmH7yG3CjyKSqZY2LyD4  fuse_replace_into_number_accumulate_merge_action : 35903.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] KyVJW8dsE1wYbkiepZrPA4  fuse_replace_into_number_apply_deletion : 4328.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] fYzUEb3BGIZfAcmlvT1FA1  fuse_replace_into_number_apply_deletion : 1470.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] JQQmH7yG3CjyKSqZY2LyD4  fuse_replace_into_number_apply_deletion : 14231.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] KyVJW8dsE1wYbkiepZrPA4  fuse_replace_into_partition_number : 13482774.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] fYzUEb3BGIZfAcmlvT1FA1  fuse_replace_into_partition_number : 16660260.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] JQQmH7yG3CjyKSqZY2LyD4  fuse_replace_into_partition_number : 16660260.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] KyVJW8dsE1wYbkiepZrPA4  fuse_replace_into_row_number_after_pruning : 11719470790.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] fYzUEb3BGIZfAcmlvT1FA1  fuse_replace_into_row_number_after_pruning : 6094381660.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] JQQmH7yG3CjyKSqZY2LyD4  fuse_replace_into_row_number_after_pruning : 29244447168.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] KyVJW8dsE1wYbkiepZrPA4  fuse_replace_into_row_number_after_table_level_pruning : 13484951.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] fYzUEb3BGIZfAcmlvT1FA1  fuse_replace_into_row_number_after_table_level_pruning : 16662834.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] JQQmH7yG3CjyKSqZY2LyD4  fuse_replace_into_row_number_after_table_level_pruning : 16662834.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] KyVJW8dsE1wYbkiepZrPA4  fuse_replace_into_row_number_source_block : 13485000.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] fYzUEb3BGIZfAcmlvT1FA1  fuse_replace_into_row_number_source_block : 16663000.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] JQQmH7yG3CjyKSqZY2LyD4  fuse_replace_into_row_number_source_block : 16663000.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] KyVJW8dsE1wYbkiepZrPA4  fuse_replace_into_row_number_totally_loaded : 607752660.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] fYzUEb3BGIZfAcmlvT1FA1  fuse_replace_into_row_number_totally_loaded : 353914055.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] JQQmH7yG3CjyKSqZY2LyD4  fuse_replace_into_row_number_totally_loaded : 2811621848.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] KyVJW8dsE1wYbkiepZrPA4  fuse_replace_into_row_number_write : 607291502.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] fYzUEb3BGIZfAcmlvT1FA1  fuse_replace_into_row_number_write : 353705055.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] JQQmH7yG3CjyKSqZY2LyD4  fuse_replace_into_row_number_write : 2808425860.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] KyVJW8dsE1wYbkiepZrPA4  fuse_replace_into_segment_number_after_pruning : 4327.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] fYzUEb3BGIZfAcmlvT1FA1  fuse_replace_into_segment_number_after_pruning : 1470.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] JQQmH7yG3CjyKSqZY2LyD4  fuse_replace_into_segment_number_after_pruning : 14219.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] KyVJW8dsE1wYbkiepZrPA4  fuse_replace_into_time_accumulated_merge_action_ms : 14503.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] fYzUEb3BGIZfAcmlvT1FA1  fuse_replace_into_time_accumulated_merge_action_ms : 8123.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] JQQmH7yG3CjyKSqZY2LyD4  fuse_replace_into_time_accumulated_merge_action_ms : 54011.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] KyVJW8dsE1wYbkiepZrPA4  fuse_replace_into_time_apply_deletion_ms : 646281.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] fYzUEb3BGIZfAcmlvT1FA1  fuse_replace_into_time_apply_deletion_ms : 378893.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] JQQmH7yG3CjyKSqZY2LyD4  fuse_replace_into_time_apply_deletion_ms : 2998191.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] JQQmH7yG3CjyKSqZY2LyD4  fuse_replace_into_time_process_input_block_ms : 15305.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] KyVJW8dsE1wYbkiepZrPA4  fuse_replace_into_time_process_input_block_ms : 14567.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] fYzUEb3BGIZfAcmlvT1FA1  fuse_replace_into_time_process_input_block_ms : 14484.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] KyVJW8dsE1wYbkiepZrPA4  replace_into_time_execution_ms : 4834740.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] KyVJW8dsE1wYbkiepZrPA4  replace_into_time_mutation_ms : 3670031.0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] ===========================
[2023-08-10T14:30:03Z INFO  test_replace_recluster]
[2023-08-10T14:30:03Z INFO  test_replace_recluster]
[2023-08-10T14:30:03Z INFO  test_replace_recluster] ======CLUSTERING INFO======
[2023-08-10T14:30:03Z INFO  test_replace_recluster] cluster_key : (to_yyyymmdd(insert_time), id)
[2023-08-10T14:30:03Z INFO  test_replace_recluster] block_count: 56
[2023-08-10T14:30:03Z INFO  test_replace_recluster] constant_block_count: 2
[2023-08-10T14:30:03Z INFO  test_replace_recluster] unclustered_block_count: 0
[2023-08-10T14:30:03Z INFO  test_replace_recluster] average_overlaps: 15.6429
[2023-08-10T14:30:03Z INFO  test_replace_recluster] average_depth: 10.2679
[2023-08-10T14:30:03Z INFO  test_replace_recluster] block_depth_histogram: {"00002":1,"00008":4,"00009":10,"00010":17,"00011":7,"00012":17}
[2023-08-10T14:30:03Z INFO  test_replace_recluster] ===========================

~~~

**Review by dantengsky — APPROVED** [review]
let's merge

**Pull request merged**